File: ios/GradeThread/Marketplaces/Reconciliation/ReconciliationTypes.swift
import Foundation

/// One unmatched eBay listing row pulled from `flipdesk_ebay_listings`.
/// Each row needs the user to choose Create-new / Link-existing / Ignore
/// before it disappears from the queue.
public struct OrphanEbayListing: Decodable, Identifiable, Equatable {
    public let id: String
    public let ebayItemId: String
    /// eBay's "Custom label" field — typically the seller's SKU. May be
    /// nil when the user hasn't set one on the listing.
    public let customLabel: String?
    public let title: String?
    public let currentPrice: Double?
    public let availableQuantity: Int?
    public let listingURL: String?
    public let listingFormat: String?
    public let importedAt: String

    private enum CodingKeys: String, CodingKey {
        case id
        case ebayItemId = "ebay_item_id"
        case customLabel = "custom_label"
        case title
        case currentPrice = "current_price"
        case availableQuantity = "available_quantity"
        case listingURL = "listing_url"
        case listingFormat = "listing_format"
        case importedAt = "imported_at"
    }

    /// Title to display in the queue row — falls back to "Listing
    /// {ebay_item_id}" when the eBay row didn't capture one.
    public var displayTitle: String {
        if let title, !title.isEmpty { return title }
        return "Listing \(ebayItemId)"
    }

    /// Default title pre-filled into the Create sheet. Same fallback as
    /// `displayTitle` but exposed under a distinct name so the intent at
    /// call sites is obvious.
    public var suggestedTitle: String { displayTitle }
}

/// Outcome of a single match attempt (Create / Link / Ignore). The
/// reconciliation view aggregates these for the bulk path.
public struct ReconciliationOutcome: Equatable {
    public let orphanId: String
    public let kind: Kind

    public enum Kind: Equatable {
        case created(itemId: String)
        case linked(itemId: String)
        case ignored
        case failed(message: String)
    }

    public var isSuccess: Bool {
        switch kind {
        case .created, .linked, .ignored: return true
        case .failed:                     return false
        }
    }
}

/// Aggregated result of a bulk Create-All run. Mirrors
/// `BulkActionResult.summary` semantics for the toast.
public struct ReconciliationBulkResult: Equatable {
    public let succeeded: Int
    public let failures: [(orphanId: String, message: String)]

    public init(succeeded: Int, failures: [(String, String)]) {
        self.succeeded = succeeded
        self.failures = failures
    }

    public var summary: String {
        let total = succeeded + failures.count
        let unit = total == 1 ? "item" : "items"
        if failures.isEmpty { return "Created \(succeeded) \(unit) from eBay." }
        if succeeded == 0 { return "All \(total) \(unit) failed." }
        return "Created \(succeeded) of \(total) \(unit); \(failures.count) failed."
    }

    /// Equatable on a typealias around the tuple — Swift doesn't
    /// synthesize Equatable for tuples in struct properties.
    public static func == (lhs: ReconciliationBulkResult, rhs: ReconciliationBulkResult) -> Bool {
        guard lhs.succeeded == rhs.succeeded,
              lhs.failures.count == rhs.failures.count else { return false }
        for (a, b) in zip(lhs.failures, rhs.failures) where a.0 != b.0 || a.1 != b.1 {
            return false
        }
        return true
    }
}

File: android/app/src/main/java/com/gradethread/app/marketplaces/reconciliation/ReconciliationTypes.kt
package com.gradethread.app.marketplaces.reconciliation

import com.gradethread.app.ui.UiMessage

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

/**
 * US-1356: one unmatched eBay listing from `flipdesk_ebay_listings`.
 *
 * Each row waits on a decision — create a new item, link it to an existing
 * one, or ignore it — before it leaves the queue.
 */
@Serializable
data class OrphanEbayListing(
    val id: String = "",
    @SerialName("ebay_item_id") val ebayItemId: String = "",
    /** eBay's "Custom label" — usually the seller's SKU. Often unset. */
    @SerialName("custom_label") val customLabel: String? = null,
    val title: String? = null,
    @SerialName("current_price") val currentPrice: Double? = null,
    @SerialName("available_quantity") val availableQuantity: Int? = null,
    @SerialName("listing_url") val listingUrl: String? = null,
    @SerialName("listing_format") val listingFormat: String? = null,
    @SerialName("imported_at") val importedAt: String? = null,
) {
    /**
     * What the queue row shows. eBay doesn't always give us a title, and
     * "Listing 12345" is at least true — a blank row is not.
     */
    val displayTitle: String
        get() = title?.takeIf { it.isNotBlank() } ?: "Listing $ebayItemId"

    /** The title pre-filled into the create sheet. Same fallback, named for intent. */
    val suggestedTitle: String get() = displayTitle
}

/** What happened to one orphan. */
sealed interface ReconcileOutcome {
    val orphanId: String

    data class Created(override val orphanId: String, val itemId: String) : ReconcileOutcome
    data class Linked(override val orphanId: String, val itemId: String) : ReconcileOutcome
    data class Ignored(override val orphanId: String) : ReconcileOutcome
    data class Failed(override val orphanId: String, val message: UiMessage) : ReconcileOutcome

    val succeeded: Boolean get() = this !is Failed
}

/** Aggregate of a create-all run. */
data class ReconcileBulkResult(val succeeded: Int = 0, val failures: List<Pair<String, UiMessage>> = emptyList()) {
    val total: Int get() = succeeded + failures.size

    /**
     * The toast line. A partial run says so — "created 8 of 10" leaves the
     * seller looking for the two that didn't, which "created 8" would hide.
     */
    val summary: String
        get() {
            val unit = if (total == 1) "item" else "items"
            return when {
                total == 0 -> "Nothing to create."
                failures.isEmpty() -> "Created $succeeded $unit from eBay."
                succeeded == 0 -> "All $total $unit failed."
                else -> "Created $succeeded of $total $unit; ${failures.size} failed."
            }
        }

    companion object {
        fun from(outcomes: List<ReconcileOutcome>): ReconcileBulkResult = ReconcileBulkResult(
            succeeded = outcomes.count { it.succeeded },
            failures = outcomes.filterIsInstance<ReconcileOutcome.Failed>()
                .map { it.orphanId to it.message },
        )
    }
}

/**
 * US-1356: how many listings are waiting, and whether that number is exact.
 *
 * [atLeast] exists because the count is read with a ceiling — "200+" is what we
 * actually know, and rendering a precise-looking 200 would claim a measurement
 * that wasn't taken.
 */
data class OrphanCount(val value: Int = 0, val atLeast: Boolean = false) {
    val isEmpty: Boolean get() = value <= 0

    /** "3 unmatched eBay listings" / "200+ unmatched eBay listings". */
    val label: String
        get() {
            val number = if (atLeast) "$value+" else "$value"
            val noun = if (value == 1 && !atLeast) "listing" else "listings"
            return "$number unmatched eBay $noun"
        }
}
